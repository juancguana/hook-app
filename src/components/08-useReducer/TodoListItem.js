import React from 'react';

const TodoListItem = ({ todo, index, handleToggle, handleDelete }) => {
  return (
    <li key={todo.id} className='list-group-item'>
      <p
        className={`${todo.done && 'complete'}`}
        onClick={() => handleToggle(todo.id)}
      >{`${index + 1}. ${todo.desc}`}</p>
      <button className='btn btn-danger' onClick={() => handleDelete(todo.id)}>
        Borrar
      </button>
    </li>
  );
};

export default TodoListItem;
