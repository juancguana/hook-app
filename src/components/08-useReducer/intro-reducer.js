const initialState = [{ id: 1, todo: 'Comprar pan', done: false }];
const ADD = 'ADD';

const todoReducer = (state = initialState, action) => {
  if (action?.type === ADD) {
    return [...state, action.payload];
  }
  return state;
};

let todos = todoReducer();

const newTodo = { id: 2, todo: 'Comprar leche', done: false };

const addTodoAction = {
  type: ADD,
  payload: newTodo,
};

todos = todoReducer(todos, addTodoAction);
console.log(todos);
