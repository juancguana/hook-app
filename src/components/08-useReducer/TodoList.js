import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul>
      {todos.map((todo, index) => (
        <TodoListItem
          todo={todo}
          key={todo.id}
          index={index}
          handleDelete={handleDelete}
          handleToggle={handleToggle}
        />
      ))}
    </ul>
  );
};
