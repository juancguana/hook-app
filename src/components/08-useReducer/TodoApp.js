import React, { useEffect, useReducer } from 'react';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';
import './styles.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    const action = {
      type: 'delete',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    dispatch({
      type: 'toggle',
      payload: todoId,
    });
  };

  const handleAddTodo = (newTodo) => {
    dispatch({
      type: 'add',
      payload: newTodo,
    });
  };

  return (
    <div>
      <h1>Todo App ({todos.length})</h1>
      <hr />
      <div className='row'>
        <div className='col-7'>
          <TodoList
            todos={todos}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
          />
        </div>
        <div className='col-5'>
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </div>
  );
};
