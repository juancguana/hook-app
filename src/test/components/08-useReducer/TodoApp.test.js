import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';

describe('first', () => {
  const wrapper = shallow(<TodoApp />);

  test('should show correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should add a todo', () => {
    const wrapper = mount(<TodoApp />);

    Storage.prototype.setItem = jest.fn(() => {});

    act(() => {
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
    });
    expect(wrapper.find('h1').text().trim()).toBe('Todo App (2)');

    expect(localStorage.setItem).toHaveBeenCalledTimes(1);
  });

  test('should delete a todo', () => {
    wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
    wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

    expect(wrapper.find('h1').text().trim()).toBe('Todo App (0)');
  });
});
