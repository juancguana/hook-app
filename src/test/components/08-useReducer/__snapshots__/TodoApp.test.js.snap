// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`first should show correctly 1`] = `
<div>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
